### PR TITLE
Feature/update tag value

### PR DESCRIPTION
**Pull request by AlcibiadesCleinias**
Hi, 

there was a discussion with possible implementation: https://github.com/Hugovdberg/PIconnect/issues/498.
Then, @ldariva contribute with pr: https://github.com/Hugovdberg/PIconnect/pull/540/files but it still is opened and with some changes suggested.

- so, my pr fully relies on the pr mentioned above
- In a code, I allowed myself to leave some function argument annotations. I believe with annotations it is easier to understand what should be passed to a method and IDE shows hints, and pep8... (until the last I have tried to follow the main rule of pep8: cohesive module style, but I guess with such annotations it is rly more comfortable and readable)
- with the new update method, it is possible to change old values with the right combination of the method options
- rm todo in the readme and wrote features list instead (not sure if completed)

Acknowledgment
- @ldariva
- @Hugovdberg 

PS.
It is better to review and merge after https://github.com/Hugovdberg/PIconnect/pull/572

### CODE REVIEW

**Comment by AlcibiadesCleinias** [comment]
M-m, @Hugovdberg I saw your use of `int(IntEnum)`. Guess, I will change behaviour to yours, i.e. instead of `IntEnum.value` -> `int(IntEnum)`. So leave the possibility to pass to a method integers

**Comment by Hugovdberg** [comment]
I chose this implementation to allow for direct use of `int` values indeed, which I believe was mostly to accommodate for versions of python.

File: PIconnect/PIData.py [hunk 1]
```diff
@@ -187,6 +195,31 @@ def current_value(self):
...
+            value: value type should be in cohesion with PI object or
+                it will raise PIException: [-10702] STATE Not Found
+            ddatetime (datetime, optional): it is not possible to set future value,
+                it raises PIException: [-11046] Target Date in Future.
+
+        You can combine update_mode and ddatetime to change already stored value.
+        """
+
+        if ddatetime:
+            assert ddatetime.tzinfo, "No timezone specified for the datetime passed."
```
Comment (Hugovdberg) [inline PIconnect/PIData.py:217]: Could you please change this to an `if ... raise` construction, raising a ValueError with that message? Codacy warns that `assert` statement can be removed if someone compiles the package to optimised byte code.

File: PIconnect/PIAF.py [hunk 2]
```diff
@@ -218,6 +218,13 @@ def units_of_measurement(self):
     def _current_value(self):
         return self.attribute.GetValue().Value
 
+    def _update_value(self, af_value_obj, update_mode, buffer_mode):
```
Comment (Hugovdberg) [inline PIconnect/PIAF.py:221]: I would prefer to call the argument just `value`, when we deprecate support for python 2 I want to add type hints to all code more explicitly.

File: PIconnect/PI.py [hunk 3]
```diff
@@ -223,6 +234,9 @@ def _current_value(self):
         """Return the last recorded value for this PI Point (internal use only)."""
         return self.pi_point.CurrentValue().Value
 
+    def _update_value(self, af_value_obj, update_mode, buffer_mode):
```
Comment (Hugovdberg) [inline PIconnect/PI.py:237]: I would prefer to call the argument just `value`, when we deprecate support for python 2 I want to add type hints to all code more explicitly.

File: PIconnect/PIData.py [hunk 4]
```diff
@@ -187,6 +195,31 @@ def current_value(self):
         Return the current value of the attribute."""
         return self._current_value()
 
+    def update_value(
+        self,
+        value,
+        ddatetime: Optional[datetime] = None,
```
Comment (Hugovdberg) [inline PIconnect/PIData.py:201]: I think `timestamp` would be a better name for this parameter without conflicting with the `datetime` class name.

File: PIconnect/PIData.py [hunk 5]
```diff
@@ -187,6 +195,31 @@ def current_value(self):
...
+        update_mode: UpdateMode = UpdateMode.NO_REPLACE,
+        buffer_mode: BufferMode = BufferMode.BUFFER_IF_POSSIBLE,
```
Comment (Hugovdberg) [inline PIconnect/PIData.py:203]: Could you remove the type hints for now so we remain python 2 compatible for the upcoming release?

File: PIconnect/PIData.py (hunk 4 above)
Comment (Hugovdberg) [reply PIconnect/PIData.py:201]: this should then also be reflected in the tutorial.

File: docs/tutorials/update_value.rst [hunk 6]
```diff
@@ -0,0 +1,25 @@
+##############################
+Update value in PI
```
Comment (Hugovdberg) [inline docs/tutorials/update_value.rst:2]: Thanks for adding an explanatory tutorial! Looks good to me!

**Review by Hugovdberg — CHANGES_REQUESTED** [review]
Thanks a lot for this contribution. This is a valuable addition to the package, as it was a widely requested feature. I have added a few comments for things I would like to see slightly different before merging. Also, don't forget to add yourself to the AUTHORS.rst as a contributor!

File: PIconnect/PIData.py (hunk 1 above)
Comment (AlcibiadesCleinias) [reply PIconnect/PIData.py:217]: hm-m, I removed assert. No raise Exception instead. 

With my PI Server, it merely writes a value with datetime of UTC

File: PIconnect/PIAF.py (hunk 2 above)
Comment (AlcibiadesCleinias) [reply PIconnect/PIAF.py:221]: np

File: PIconnect/PI.py (hunk 3 above)
Comment (AlcibiadesCleinias) [reply PIconnect/PI.py:237]: np

File: PIconnect/PIData.py (hunk 4 above)
Comment (AlcibiadesCleinias) [reply PIconnect/PIData.py:201]: ok, what I have thought about when I tried to choose b/w:
1. time - this name represents only time, i.e. without date info, thus as a name for instance that shows time and date simultaneously it is not suited well.
2. ddatetime - it came as a kinda compromise b/w 1 point and already existed datetime type
3. timespan - it is not suited even more than in the 1 point

What should we do?
I guess, let's stop on simple "time":
- you use it in methods: recorded_values(..., start_time, ...), interpolated_values(..., start_time, ...). So, we will follow module`s style
- the 1 point mentioned above is questioned more than 3d, so... 
- ddatetime looks like a pitfall, dt is a more correct one. But, follow module style is better

File: PIconnect/PIData.py (hunk 5 above)
Comment (AlcibiadesCleinias) [reply PIconnect/PIData.py:203]: okies

File: docs/tutorials/update_value.rst (hunk 6 above)
Comment (AlcibiadesCleinias) [reply docs/tutorials/update_value.rst:2]: thanks to @ldariva

**Review by Hugovdberg — APPROVED** [review]

**Pull request merged**